controller/src/lib.rs

```diff
@@ -20,114 +21,168 @@ pub mod plan;
+            Ok(ScheduleResponse::NoDroneAvailable)
+        }
+    }
+}
+
+/// get backend associated with a lock or error
+fn backend_of_lock(
+    state: &StateHandle,
+    cluster_name: &ClusterName,
+    lock: &str,
+) -> anyhow::Result<BackendId> {
+    state
+        .state()
+        .cluster(cluster_name)
+        .ok_or_else(|| anyhow!("no cluster"))?
+        .locked(lock)
+        .ok_or_else(|| anyhow!("no backend"))
+}
+
+fn schedule_response_for_existing_backend(
+    state: &StateHandle,
+    cluster: ClusterName,
+    backend: BackendId,
+) -> anyhow::Result<ScheduleResponse> {
+    // Anything that fails to find the drone results in an error here, since we just
+    // checked that the lock is held which implies that the drone exists.
+    let state = state.state();
+    tracing::info!("getting cluster from state");
+    let cluster = state
+        .cluster(&cluster)
+        .ok_or_else(|| anyhow!("no such cluster"))?;
+    tracing::info!("fetching backend!");
+    let (drone, bearer_token) = {
+        let backend_state = cluster
+            .backend(&backend)
+            .ok_or_else(|| anyhow!("Lock held by a backend that doesn't exist."))?;
+
+        let drone = backend_state
+            .drone
+            .clone()
+            .ok_or_else(|| anyhow!("Lock held by a backend without a drone assignment."))?;
+
+        let bearer_token = backend_state.bearer_token.clone();
+
+        (drone, bearer_token)
+    };
+
+    Ok(ScheduleResponse::Scheduled {
+        drone,
+        backend_id: backend,
+        bearer_token,
+        spawned: false,
+    })
+}
+
+async fn process_response(
+    state: &StateHandle,
+    sr: &ScheduleRequest,
+    scheduler: &Scheduler,
+    nats: &TypedNats,
+) -> anyhow::Result<ScheduleResponse> {
+    tracing::info!("checking locks");
+    let cluster_name = sr.cluster.clone();
+    if let Some(lock_name) = sr.lock.clone() {
+        tracing::info!(?lock_name, "scheduling lock");
+
+        if let Ok(backend) = backend_of_lock(state, &cluster_name, &lock_name) {
+            tracing::info!(?backend, "fetch preexisting backend");
+            return schedule_response_for_existing_backend(state, cluster_name, backend);
+        }
+
+        tracing::info!("spawn with lock");
+    }
+
+    match scheduler.schedule(&cluster_name, Utc::now()) {
+        Ok(drone_id) => spawn_backend(nats, &drone_id, &sr.clone()).await,
+        Err(SchedulerError::NoDroneAvailable) => Ok(ScheduleResponse::NoDroneAvailable),
+    }
+}
+
+async fn dispatch_schedule_request(
+    state: StateHandle,
+    schedule_request: MessageWithResponseHandle<ScheduleRequest>,
+    scheduler: Scheduler,
+    nats: TypedNats,
+) -> anyhow::Result<()> {
+    let Ok(response) = process_response(
+		&state,
+		&schedule_request.value.clone(),
+		&scheduler,
+		&nats
+	).await else {
+		tracing::error!(?schedule_request.value, "schedule request failed");
+		panic!("failed to dispatch");
```

Comment:
Rather than panic here, let's just exit. The panic is lost anyway and we're already logging a message that will be more relevant than the panic message, so it's less noisy just to exit.